### PR TITLE
Fix for TokenEater force logout exploit

### PR DESCRIPTION
**Pull request by Dex472**
Fix for the exploit https://github.com/Dex472/TokenEater

### CODE REVIEW

File: http_server/www/logout.php [hunk 1]
```diff
@@ -5,13 +5,33 @@
 
 header("Content-type: text/plain");
 
+function is_from_game(){
```
Comment (bls1999) [inline http_server/www/logout.php]: The formatting is off.

File: http_server/www/logout.php [hunk 2]
```diff
@@ -5,13 +5,33 @@
...
+  {
+      return false;   
+  }
+ }
+ else
+ {
+    return false;
+ }
+}
+
```
Comment (bls1999) [inline http_server/www/logout.php]: The thought of using `HTTP_X_REQUESTED_WITH` crossed my mind; However, this too can be easily spoofed.

File: http_server/www/logout.php [hunk 3]
```diff
@@ -5,13 +5,33 @@
...
 $ip = get_ip();
 
 try {
     // rate limiting
     rate_limit('logout-'.$ip, 5, 2, 'Please wait at least 5 seconds before attempting to log out again.');
     rate_limit('logout-'.$ip, 60, 10, 'Only 10 logout requests per minute per IP are accepted.');
 
+    if (is_from_game() !== true){
+     throw new Exception("For security reasons logout can only be performed from game");
```
Comment (bls1999) [inline http_server/www/logout.php]: Change language to keep consistent with other third-party site warnings. i.e. "It looks like you're not using PR2 to log out. For security reasons, you may only log out from a PR2 client."

**Comment by jacob-grahn** [comment]
Does flash always include the HTTP_X_REQUESTED_WITH header with outbound requests?

**Comment by Dex472** [comment]
@jacob-grahn Just checked it and it does send it everytime the request's origin is from the pr2 client.

**Comment by jacob-grahn** [comment]
:+1:

**Pull request merged**